example/poisson/poisson_package.cpp

```diff
@@ -49,59 +50,123 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+        // first add the RHS
+        sum += v(b, irho, k, j, i) * std::pow(dx, ndim);
+      },
+      Kokkos::Sum<Real>(total));
+
+  *reduce_sum += total;
+  return TaskStatus::complete;
+}
+
+template <typename T>
+TaskStatus ComputeRHS(T *u) {
+  auto pm = u->GetParentPointer();
+
+  IndexRange ib = u->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = u->GetBoundsK(IndexDomain::interior);
+
+  PackIndexMap imap;
+  const std::vector<std::string> vars({"density", "rhs"});
+  const auto &v = u->PackVariables(vars, imap);
+  const int irho = imap["density"].first;
+  const int irhs = imap["rhs"].first;
+
+  auto coords = GetCoords(pm);
+  const int ndim = v.GetNdim();
+  const Real dx = coords.Dx(X1DIR);
+  for (int i = X2DIR; i <= ndim; i++) {
+    const Real dy = coords.Dx(i);
+    PARTHENON_REQUIRE_THROWS(dx == dy,
+                             "ComputeRHS requires that DX be equal in all directions.");
   }
 
   parthenon::par_for(
-      DEFAULT_LOOP_PATTERN, "UpdatePhi", DevExecSpace(), 0, v.GetDim(5) - 1, kb.s, kb.e,
+      DEFAULT_LOOP_PATTERN, "ComputeRHS", DevExecSpace(), 0, v.GetDim(5) - 1, kb.s, kb.e,
       jb.s, jb.e, ib.s, ib.e,
       KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
-        dv(b, idphi, k, j, i) = -v(b, irho, k, j, i) * std::pow(dx, ndim);
-        dv(b, idphi, k, j, i) += v(b, iphi, k, j, i - 1) + v(b, iphi, k, j, i + 1);
-        if (ndim > 1) {
-          dv(b, idphi, k, j, i) += v(b, iphi, k, j - 1, i) + v(b, iphi, k, j + 1, i);
-          if (ndim == 3) {
-            dv(b, idphi, k, j, i) += v(b, iphi, k - 1, j, i) + v(b, iphi, k - 1, j, i);
-          }
-        }
-        dv(b, idphi, k, j, i) /= 2.0 * ndim;
-        dv(b, idphi, k, j, i) -= v(b, iphi, k, j, i);
+        // first add the RHS
+        v(b, irhs, k, j, i) = -v(b, irho, k, j, i) * std::pow(dx, ndim);
       });
+  return TaskStatus::complete;
+}
+
+template <typename T>
+TaskStatus UpdatePhi(T *u, T *du) {
+  using Stencil_t = parthenon::solvers::Stencil<Real>;
+  Kokkos::Profiling::pushRegion("Task_Poisson_UpdatePhi");
+  auto pm = u->GetParentPointer();
+
+  IndexRange ib = u->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = u->GetBoundsK(IndexDomain::interior);
+
+  PackIndexMap imap;
+  const std::vector<std::string> vars({"rhs", "potential"});
+  const auto &v = u->PackVariables(vars, imap);
+  const int irhs = imap["rhs"].first;
+  const int iphi = imap["potential"].first;
+  const std::vector<std::string> phi_var({"potential"});
+  PackIndexMap imap2;
+  const auto &dv = du->PackVariables(phi_var, imap2);
+  const int idphi = imap2["potential"].first;
+
+  StateDescriptor *pkg = pm->packages.Get("poisson_package").get();
+  const auto stencil = pkg->Param<Stencil_t>("stencil");
+
+  StencilMatVec(v, iphi, dv, idphi, v, irhs, stencil, ib, jb, kb);
```

Comment:
Written in this way, you introduce an extra kernel launch. I wonder if we could add a `Stencil` call that computes and fills `dv` and does the update all at once?

Reply:
Hmm. I think naming things similar to blas for continuity is probably useful.

But you're right for now we should focus on the need we have.